optimization/eda/probability-learning.metta

```diff
@@ -0,0 +1,291 @@
+;; ================================================================================
+;; Probability Learning Policy
+;; ================================================================================
+;;
+;; This file implements the probability learning policy for EDA.
+;; Equivalent to the model building logic in local_structure.h/cc in the C++ implementation.
+;;
+;; Key Components:
+;; - Learn probability distributions from selected instances
+;; - Count knob value occurrences
+;; - Compute probabilities with Laplace smoothing
+;; - Normalize probability distributions
+;;
+;; Data Structures:
+;; - KnobProbabilityModel: mkKnobProbabilityModel (Map KnobId (Map KnobValue Number))
+;; - KnobValue: mkKnobValue Number
+;; - KnobId: NodeId
+
+;; ================================================================================
+;; Type Definitions
+;; ================================================================================
+
+;; Knob Value and Probability Model
+(: KnobValue Type)
+(: mkKnobValue (-> Number KnobValue))
+(: KnobId Type)
+(: KnobProbabilityModel Type)
+(: mkKnobProbabilityModel (-> (Map KnobId (Map KnobValue Number)) (KnobProbabilityModel)))
+
+;; ================================================================================
+;; Probability Learning Policy
+;; ================================================================================
+
+;; Learn probability distributions from selected instances
+; (: learn-probability-model (-> Deme (InstanceSet $score) (KnobProbabilityModel)))
+(= (learn-probability-model $deme $selectedInstances)
+  (let* (
+        (() (println! (learn-probability-model: starting)))
+        ((mkDeme $rep (mkSInstSet $instSet) $id) $deme)
+        (() (println! (learn-probability-model: got rep $rep)))
+        ($discMap (getDiscMap $rep))
+        (() (println! (learn-probability-model: got discMap $discMap)))
+        ($probModel (learn-knob-probabilities $discMap $selectedInstances))
+        (() (println! (learn-probability-model: got probModel $probModel)))
+      )
+      $probModel))
+
+;; Learn probability distributions for each knob from selected instances
+; (: learn-knob-probabilities (-> (MultiMap (DiscSpec $knob)) (InstanceSet $score) (KnobProbabilityModel)))
+(= (learn-knob-probabilities $discMap $selectedInstances)
+  (let* (
+        (() (println! (learn-knob-probabilities: starting)))
+        (() (println! (learn-knob-probabilities: got discMap $discMap)))
+        (() (println! (learn-knob-probabilities: got selectedInstances $selectedInstances)))
+        ($knobCounts (count-knob-values $discMap $selectedInstances))
+        (() (println! (learn-knob-probabilities: got knobCounts $knobCounts)))
+        ($probabilities (compute-probabilities $knobCounts))
+        (() (println! (learn-knob-probabilities: got probabilities $probabilities)))
+      )
+      (mkKnobProbabilityModel $probabilities)))
+
+;; Count occurrences of each knob value in selected instances
```

Comment:
Can you use higher order functions and / or any other method to reduce the complexity of these implementations? I couldn't understand or keep up with the level of nested functions. It will also be really hard to maintain a code that you can't understand.